rfcs/20230118-f16-accumulation/README.md

```diff
@@ -0,0 +1,125 @@
+accumulation datatype in oneDNN is f32, for floating point
+computations, and s32 for integer computation.
+
+Some platforms allow additional speedups when using smaller
+accumulation datatype.  This is in particular the case for f16
+primitives, where up to 2x speedups can be obtained by using f16
+accumulation datatype on some platforms.
+
+# 2. Problem statement
+
+As of now, oneDNN officially supports only inference for f16 models,
+and many data show that good levels of accuracy can be maintained
+while using f16 accumulation datatype for most f16 models. We hence
+have been dispatching implementations with f16 accumulation on some
+platforms.
+
+Even though dispatching implementations with f16 accumulation is
+acceptable for inference, it is not for training, which requires the
+full f32 precision accumulators. Hence, as oneDNN is planning
+support for f16 training, it is important that we dispatch only
+implementation with f32 accumulation for training workloads.
+
+Note that the main user for oneDNN f16 implementations is openVINO,
+which is the requester for the implementations with f16 accumulation
+in the first place.
+
+
+# 3. Options
+
+## Using the propagation kind
+
+One natural option would be to rely on propagation kind.  If the
+propagation kind, is `forward_inference`, we allow dispatching of
+implementations with f16 accumulation, otherwise we don't.  This works
+for most primitives except for Matmul, which does not have a
+propagation kind. Hence this option does not cover the whole library.
+
+## Using the `ONEDNN_ENABLE_WORKLOAD` build option
+
+oneDNN currently has a `ONEDNN_ENABLE_WORKLOAD` build option that can
+be set to `INFERENCE` or `TRAINING` (default). This option is
+currently used to disable dispatching of some primitive
+implementations based on propagation_kind. We could rely on this
+mecanism to guard dispatching of implementations using f16
+accumulation with the two following caveats:
+- this would increase validation costs as we would need to validate both builds
+- it adds development burden. Currently we don't know any information
+  on the accumulation datatype for each implementations. Properly
+  guarding implementation dispatch would either need to be scattered
+  in implementation using f16 accumulation (which is brittle), or
+  would require to add internal APIs to query implementations for their
+  accumulation datatype.
+
+## A new `accumulation_datatype` attribute (recommended)
+
+The last option is to rely on a new primitive attribute to get a hint
+on minimal accuracy required in accumulators. There are multiple
+options for what values it should accept.
+
+For all those options, note that implementations would still be able
+to use a different datatype internally if the result is unchanged and
+it is faster (e.g. using `f32` intermediate accumulators when
+accumulation datatype is `f16`, or in some `s8/u8` computations when
+it yields the same result).
```

Comment:
Would it be reasonable to make this a modifcation to `fpmath-mode`? From our description of fpmath-mode:

> When passed to a primitive creation, the [dnnl::fpmath_mode](https://oneapi-src.github.io/oneDNN/enum_dnnl_fpmath_mode.html#doxid-group-dnnl-api-fpmath-mode-1ga0ad94cbef13dce222933422bfdcfa725) primitive attribute specifies which implicit down-conversions are allowed for that given primitive. Only down-conversions from f32 to narrower data-types (f16, bf16, or tf32) are currently allowed. 

We never specify what can be downcoverted, so an `f16` workload with `fpmath-mode=f16` could be interpreted to enable down-coverting f32 accumulation to `f16`.

Reply:
fpmath-mode requires f32 inputs to have an effect, which doesn't help the general case considered in the RFC. Applying this API for non-f32 inputs brings rather confusion than a solution. And if by some reason, user would like to convert f32 into f16 implicitly and utilize f16 accumulators, it's not clear how using this API will help them to define an accumulator data type because by default it's f32 and using fpmath=f16 brings clear intention to down-convert inputs only but not accumulator data type.

Reply:
@dzarukin, I think I was expounding upon that exact issue in an edit at the same time as you made your comment. Let me know what you think. I agree it is a potential issue. Just to expound on an implementation point of view: at its core, fpmath-mode is about enabling an implementation to use the most appropriate instruction for execution. Changing the accumulation instruction targets the exact same task. Why add `math-modes x accumlators` API states for choosing the instruction when there are only a few instructions modes we will ever consider targeting.  It seems simpler to just enumerated the desired modes explicitly.

Reply:
@rjoursler It is actually a very good point. I hesitated at first to make that part of fpmath mode but did not exactly for the reason you mentioned: fpmath mode covers implicit conversions from f32 to other data-types for inputs. If we cover the accumulators with it as well, we will have to expose multiple enum values to cover all configurations to support (e.g. convert f32 inputs but not accumulators, convert both inputs/accumulators, convert only accumulators).

I am not hugely against it, but I am afraid it would blow up over time, and separating them from now might be cleaner.

Reply:
Ok @mgouicem, thanks for you thoughts. I am fine with relying with your judgement on this. I mostly wanted to bring it up as an option to consider.